README.md
# koot-diagnose

Automate test cases for any website, webapp, HTML5 page and more. Uses by _Koot.js_ project's `koot-analyze` script. Each diagnose/test case can run programmably or through CLI.

## Diagnose/Test Cases

**lighthouse**

进行一次灯塔 (Lighthouse) 基准测试和检测

**crawler**

利用爬虫尝试访问站点内的所有链接，检测以下内容

-   损坏的链接
-   文件尺寸过大的请求 (如图片、JS 代码等)
-   没有 gzip 的请求
-   页面载入过程中的报错

**styles**

对指定页面进行检测，找出不合理的样式，并给出修改意见

**memory**

在指定站点内随机的进行操作，持续一段较长的时间，检查内存占用情况

**psi**

对目标域名的首页进行 PageSpeed Insights 检测


CHANGELOG.md
## [Unreleased]

## [0.13.0] - 2019-12-09

**psi**

-   新的检测：对目标域名的首页进行 PageSpeed Insights 检测

## [0.12.0] - 2019-11-08

-   代码库从 `koot` 中独立出来

**crawler**

-   新的选项: `similarCountMax`
    -   _number_ 或 _false_ 相似地址的最多请求次数，`false` 或`负数`表示无限制
    -   默认为 `10`


__tests__/psi.test.js
// jest configuration
jest.setTimeout(10 * 60 * 1 * 1000);

//

const psi = require('../psi');
const getReqUrl = require('../commons/get-psi-request-url');

//

describe('koot-diagnose', () => {
    test('PageSpeed Insights API', async () => {
        let err, result;

        try {
            result = await psi('https://cmcm.com');
        } catch (e) {
            console.error(e);
            err = e;
        }

        expect(typeof err).toBe('undefined');
        expect(typeof result).toBe('object');
    });

    test('有多个 category 时，拼成的 URL', () => {
        expect(getReqUrl('https://domain.com')).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com`
        );
        expect(getReqUrl('https://domain.com', 'abc')).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com&key=abc`
        );
        expect(
            getReqUrl('https://domain.com', 'abc', {
                category: 'seo'
            })
        ).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com&key=abc&category=seo`
        );
        expect(
            getReqUrl('https://domain.com', undefined, {
                category: 'seo'
            })
        ).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com&category=seo`
        );
        expect(
            getReqUrl('https://domain.com', 'abc', {
                category: ['seo', 'pwa']
            })
        ).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com&key=abc&category=seo&category=pwa`
        );
        expect(
            getReqUrl('https://domain.com', undefined, {
                category: ['seo', 'pwa']
            })
        ).toBe(
            `https://www.googleapis.com/pagespeedonline/v5/runPagespeed?url=https://domain.com&category=seo&category=pwa`
        );
    });
});


__tests__/crawler.test.js
// jest configuration
jest.setTimeout(10 * 60 * 1 * 1000);

//

const fs = require('fs-extra');
const path = require('path');
const execSync = require('child_process').exec;
const crawler = require('../crawler');
const waitForPort = require('./_libs/get-port-from-child-process');
const terminate = require('./_libs/terminate-process');

//

const spawn = async (cmd, o = {}) => {
    const chunks = cmd.split(' ');
    await new Promise(resolve => {
        const child = require('child_process').spawn(chunks.shift(), chunks, {
            stdio: 'inherit',
            shell: true,
            ...o
        });
        child.on('close', () => {
            resolve();
        });
    });
};

//

describe('koot-diagnose', () => {
    const dir = path.resolve(__dirname, './project-full-of-shits');

    test('crawler', async () => {
        const configFile = `koot.config.js`;
        const dist = path.resolve(dir, 'dist');
        if (fs.existsSync(dist)) fs.emptyDirSync(dist);
        else fs.removeSync(dist);

        await spawn('npm i --no-save', {
            stdio: false,
            cwd: dir
        });

        const child = execSync(`npm run start`, {
            cwd: dir
        });
        const errors = [];

        await waitForPort(child);
        const port = require(path.resolve(dir, configFile)).port;
        child.stderr.on('data', err => {
            errors.push(err);
        });

        const crawlerResult = await crawler(`http://localhost:${port}`);
        // console.log(crawlerResult);

        const {
            'console error': consoleError,
            'broken request': brokenRequest,
            'large file': largeFile
        } = crawlerResult;

        await terminate(child.pid);

        expect(typeof crawlerResult).toBe('object');
        expect(Array.isArray(crawlerResult)).toBe(false);

        expect(Array.isArray(consoleError)).toBe(true);
        expect(
            consoleError.some(
                e =>
                    typeof e === 'object' &&
                    e.message &&
                    /\/this-is-a-non-exist-css-file\.css/.test(e.message)
            )
        ).toBe(true);
        expect(
            consoleError.some(
                e => typeof e === 'object' && e.message && /404/.test(e.message)
            )
        ).toBe(true);

        expect(Array.isArray(brokenRequest)).toBe(true);
        expect(
            brokenRequest.some(
                e =>
                    typeof e === 'object' &&
                    e.url &&
                    /\/this-is-a-non-exist-js-file\.js/.test(e.url)
            )
        ).toBe(true);
        expect(
            brokenRequest.some(
                e =>
                    typeof e === 'object' &&
                    e.url &&
                    /\/this-is-a-404/.test(e.url)
            )
        ).toBe(true);

        expect(Array.isArray(largeFile)).toBe(true);
        expect(
            largeFile.some(
                e =>
                    typeof e === 'object' &&
                    e.url &&
                    /\/this-is-a-large-file\.png/.test(e.url) &&
                    e.threshold < e.contentLength
            )
        ).toBe(true);
    });
});


psi.d.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
// declare module 'kootDiagnosePSI';

export type Category =
    | 'accessibility'
    | 'best-practices'
    | 'performance'
    | 'pwa'
    | 'seo';

export interface Result {
    lighthouseResult: {
        audits: {
            [auditId: string]: Audit;
        };
        categories: {
            [categoryId: string]: {
                id: string;
                title: string;
                auditRefs: {
                    [auditId: number]: {
                        id: string;
                        weight: number;
                        group?: string;
                    };
                };
                score: number;
                description?: string;
                manualDescription?: string;
            };
        };
        categoryGroups: {
            [groupId: string]: {
                title: string;
                description?: string;
            };
        };
        [key: string]: any;
    };
    [errorType: string]: any;
}

export type AuditDetailsType =
    | 'table'
    | 'opportunity'
    | 'debugdata'
    | 'filmstrip';
export type AuditDetailsHeadingType =
    | 'text'
    | 'url'
    | 'timespanMs'
    | 'thumbnail'
    | 'bytes'
    | 'ms'
    | 'code';
export interface AuditDetailsHeading {
    key: string;
    label?: string;
    text?: string;
    valueType?: AuditDetailsHeadingType;
    itemType?: AuditDetailsHeadingType;
    displayUnit?: 'duration' | 'kb';
}

export interface Audit {
    id: string;
    title: string;
    description?: string;
    score: string;
    scoreDisplayMode: string;
    numericValue?: number;
    displayValue?: string;
    details: {
        type: AuditDetailsType;
        headings?: Array<AuditDetailsHeading>;
        items?: Array<{
            [key: string]: any;
        }>;
        overallSavingsMs?: number;
        overallSavingsBytes?: number;
        summary?: {
            [summaryKey: string]: any;
        };
        scale?: number;
        [detail: string]: any;
    };
    warnings?: Array<any>;
}

/**
 * 对目标域名的首页进行 PageSpeed Insights 检测
 */
async function psi(
    url: string | URL,
    /**
     * Google PageSpeed API Key
     * - 如果提供，检测时会使用该 Key 进行请求
     */
    apiKey?: string,
    /**
     * Google PageSpeed API 参数
     * https://developers.google.com/speed/docs/insights/v5/reference/pagespeedapi/runpagespeed#parameters
     */
    parameters: {
        category?: Category | Category[];
        locale?: string;
        strategy?: 'desktop' | 'mobile';
        utm_campaign?: string;
        utm_source?: string;
        [errorType: string]: any;
    } = {}
): Promise<Result>;

export default psi;


crawler.js
// const puppeteer = require('puppeteer');
const cheerio = require('cheerio');
const { Cluster } = require('puppeteer-cluster');

/** 忽略的 HTTP code */
const ignoredStatus = [301, 302, 304];
/** 大文件尺寸阈值 */
const largeFileThreshold = 300 * 1024; // 300KB

/**
 * 利用爬虫尝试访问站点内的所有链接，检测以下内容
 * -   损坏的链接
 * -   文件尺寸过大的请求 (如图片、JS 代码等)
 * -   HTML、CSS、JS 没有 gzip 的请求
 * -   页面载入过程中的报错
 * @async
 * @param {string} urlEntry
 * @param {Object} [options={}]
 * @param {number} [options.maxCrawl] 请求的页面数量总上限
 * @param {Object} [options.cluster] puppeteer-cluster 选项
 * @param {number} [options.similarCountMax=10] 相似地址的最多请求次数，`false` 或`负数`表示无限制
 * @param {boolean} [options.verbose=false] 是否输出更多的日志
 * @return {Promise<Object>}
 */
const crawler = async (urlEntry, options = {}) => {
    const {
        // debug = false,
        maxCrawl = undefined,
        cluster: clusterOptions = {},
        similarCountMax = 10,
        verbose = false
    } = options;
    const errors = [];
    const errObj = {};
    // let browser = await puppeteer.launch();
    // let context;
    // let page;
    let startUrl;
    let crawlCount = 0;
    const urls = {
        /** 已访问的 URL */
        visited: [],
        /** URL 队列 */
        queue: [],
        knownRoutesCount: {}
    };
    const cluster = await Cluster.launch({
        concurrency: Cluster.CONCURRENCY_CONTEXT,
        maxConcurrency: 2,
        ...clusterOptions
    });

    /**
     * 将 URL 添加至访问队列中。外站地址会被忽略
     * @param {string} url
     * @void
     */
    const addUrl = (_url, base = startUrl.origin) => {
        const url = new URL(_url, base);
        const allUrls = urls.visited.concat(urls.queue);

        // 如果已经访问过或在访问队列中，忽略
        if (allUrls.includes(url.href)) return;
        // 如果是外站地址，忽略
        if (url.origin !== startUrl.origin) return;
        // 如果相似的地址出现过多次，忽略
        {
            const segs = url.pathname.split('/');
            segs.pop();
            const pathnameExlucdeLastSeg = segs.join('/');
            if (
                pathnameExlucdeLastSeg &&
                urls.knownRoutesCount[pathnameExlucdeLastSeg]
            ) {
                const count = urls.knownRoutesCount[pathnameExlucdeLastSeg];
                if (
                    similarCountMax === false ||
                    (typeof similarCountMax === 'number' &&
                        similarCountMax < 0) ||
                    count < similarCountMax
                ) {
                    urls.knownRoutesCount[pathnameExlucdeLastSeg] = count + 1;
                } else {
                    if (verbose)
                        console.log(
                            `Path '${url.pathname}' similar to other ${similarCountMax} urls. Ignored.`
                        );
                    return;
                }
            } else {
                urls.knownRoutesCount[pathnameExlucdeLastSeg] = 1;
            }
            //     const similarCount = allUrls.reduce((count, existUrl) => {
            //         (new URL(existUrl)).pathname.split('/');
            //         return count;
            //     }, 0);
            //     if (similarCount > similarCountThreshold) return;
        }

        if (!maxCrawl || crawlCount < maxCrawl) {
            crawlCount++;
            urls.queue.push(url.href);
            cluster.queue(url.href);
        }
    };

    /**
     * 将错误添加至结果队列中
     * @param {*} type
     * @param {*} res
     * @param {*} pageUrl
     */
    const addError = (type, res, infos = {}) => {
        const { pageUrl } = infos;
        const isErrorExist = (type, msg) =>
            errors.some(e => e.message === msg && e.type === type);
        if (type instanceof Error) {
            const error = type;
            infos = res || {};
            const { pageUrl } = infos;
            const { msg, type: _type, url } = (() => {
                let matches;
                if (
                    (matches = /^net::(.+?) at (.+)$/.exec(error.message)) &&
                    matches.length &&
                    matches.length > 2
                ) {
                    return {
                        msg: `request \`${matches[2]}\` broken (net::${matches[1]})`,
                        url: matches[2],
                        type: 'broken request'
                    };
                }
                switch (error.message) {
                    default: {
                        return {
                            type: `common error`,
                            msg: error.message
                        };
                    }
                }
            })();
            if (isErrorExist(_type, msg)) return;
            errors.push(
                // Object.assign(new Error(msg), {
                //     type: _type,
                //     url,
                //     pageUrl
                // })
                {
                    message: msg,
                    type: _type,
                    url,
                    pageUrl
                }
            );
            return;
        }

        const { msg, ...r } = (() => {
            switch (type) {
                case 'broken request': {
                    return {
                        msg: `request \`${res.url()}\` broken (${res.status()} ${res.statusText()})`
                    };
                }
                case 'no gzip': {
                    return {
                        msg: `request \`${res.url()}\` should be Gzip`,
                        contentLength: infos.contentLength
                    };
                }
                case 'large file': {
                    return {
                        msg: `request \`${res.url()}\` is too large (≥ ${largeFileThreshold /
                            1024}KB)`,
                        threshold: largeFileThreshold,
                        contentLength: infos.contentLength
                    };
                }
                case 'console error': {
                    return {
                        msg: infos.consoleMsg.text(),
                        location: infos.consoleMsg.location()
                    };
                }
                default: {
                    return '';
                }
            }
        })();
        if (isErrorExist(type, msg)) return;
        const url = res ? res.url() : pageUrl || undefined;
        errors.push(
            Object.assign(
                // new Error(msg),
                {
                    message: msg,
                    type,
                    // res,
                    url,
                    pageUrl: url !== pageUrl ? pageUrl : undefined
                },
                r
            )
        );
    };

    const crawl = async (page, url = urls.queue[0]) => {
        // if (debug) console.log(`visiting: ${url}`);
        // try {
        //     await page.close();
        //     await context.close();
        // } catch (e) {}

        const checkResponse = async res => {
            if (ignoredStatus.includes(res.status())) return;

            const thisUrl = res.url();
            const pageUrl = thisUrl !== url ? url : undefined;

            if (!res.ok()) {
                return addError('broken request', res, {
                    pageUrl
                });
            }

            if (new URL(thisUrl).origin !== new URL(pageUrl || url).origin)
                return;

            const headers = res.headers();
            const contentType =
                headers['content-type'] || headers['Content-Type'];
            const contentEncoding =
                headers['content-encoding'] || headers['Content-Encoding'];
            const contentLength =
                headers['content-length'] || headers['Content-Length'];

            if (
                /^text\//.test(contentType) ||
                /\/javascript$/.test(contentType) ||
                /\/css$/.test(contentType)
            ) {
                if (!/gzip/.test(contentEncoding)) {
                    return addError('no gzip', res, {
                        pageUrl,
                        contentLength: parseInt(contentLength)
                    });
                }
            }

            if (contentLength && parseInt(contentLength) > largeFileThreshold) {
                return addError('large file', res, {
                    pageUrl,
                    contentLength: parseInt(contentLength)
                });
            }
        };

        // context = await browser.createIncognitoBrowserContext();
        try {
            // page = await context.newPage();
            const res = await page
                .on('response', checkResponse)
                .on('console', msg => {
                    if (['error', 'trace'].includes(msg.type())) {
                        addError('console error', undefined, {
                            consoleMsg: msg,
                            pageUrl: url
                        });
                    }
                })
                .goto(url, {
                    waitUntil: 'networkidle2'
                })
                .catch(e => {
                    addError(e, { pageUrl: url });
                    // return context.close();
                });

            if (!startUrl) {
                startUrl = new URL(page.url());
                url = startUrl.href;
                urls.queue.push(url);
            }

            if (res) {
                await checkResponse(res);

                const HTML = await res.text();
                const $ = cheerio.load(HTML);

                const selector = 'body a[href]';

                // add SSR links
                Object.values($(selector)).forEach(el => {
                    if (!el || !el.attribs) return;
                    try {
                        addUrl(el.attribs.href, page.url());
                    } catch (e) {
                        addError(e, { pageUrl: url });
                    }
                });

                // add CSR links
                (
                    await page.$$eval(selector, links =>
                        links
                            .map(link => {
                                if (!link) return undefined;
                                const href = link.getAttribute('href');
                                return href;
                            })
                            .filter(url => !!url)
                    )
                ).forEach(href => {
                    try {
                        addUrl(href, page.url());
                    } catch (e) {
                        addError(e, { pageUrl: url });
                    }
                });
            }

            // await page.close();
            // await context.close();

            // 将 url 从 queue 移动到 visited
            urls.visited.push(url);
            urls.queue.splice(urls.queue.indexOf(url), 1);
        } catch (e) {
            console.error(e);
            // await page.close();
            // await context.close();
        }

        // if (urls.queue.length) return await crawl(urls.queue[0]);
        // return errors;
    };

    await cluster.task(async ({ page, data: url }) => {
        await crawl(page, url);
        // const screen = await page.screenshot();
        // Store screenshot, do something else
    });

    // await crawl(urlEntry);

    crawlCount++;
    cluster.queue(urlEntry);

    // await browser.close();
    await cluster.idle();

    // console.log(urls.queue);
    // while (urls.queue.length && (!maxCrawl || crawlCount < maxCrawl)) {
    //     urls.queue.forEach(url => cluster.queue(url));
    //     crawlCount++;
    //     await cluster.idle();
    // }

    await cluster.close();

    errors.forEach(e => {
        const { type } = e;
        if (!errObj[type]) errObj[type] = [];
        errObj[type].push(e);
    });

    return errObj;
};

module.exports = crawler;


crawler.d.ts
declare module 'kootDiagnoseCrawler';

export interface CrawlerOptions {
    /** 请求的页面数量总上限 */
    maxCrawl?: number;
    /** 相似地址的最多请求次数，`false` 或`负数`表示无限制 */
    similarCountMax?: number | false;
    /**
     * 是否输出更多的日志
     * @default false
     */
    verbose?: boolean;
    /** `puppeteer-cluster` 选项 */
    cluster?: {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        [option: string]: any;
    };
}
export type CrawlerResult = {
    [errorType: string]: CrawlerError[];
};
export interface CrawlerError {
    /** 错误信息 */
    readonly message?: string;
    /** 错误类型 */
    readonly type?: string;
    /** 错误关联的 URL */
    readonly url: string;
    /** 当前页面 URL (和错误 URL 不同时) */
    readonly pageUrl?: string;
    /** 此次请求的长度 (单位: 字节) */
    readonly contentLength?: number;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
}

//

/**
 * 利用爬虫尝试访问站点内的所有链接，检测以下内容
 * -   损坏的链接
 * -   文件尺寸过大的请求 (如图片、JS 代码等)
 * -   HTML、CSS、JS 没有 gzip 的请求
 * -   页面载入过程中的报错
 */
async function crawler(
    entryURL: string,
    options?: CrawlerOptions
): Promise<CrawlerResult>;

export default crawler;


index.js
const fs = require('fs-extra');
const path = require('path');
const program = require('commander');
const chalk = require('chalk');
const { parse, stringify } = require('flatted/cjs');

//

program
    .version(require('./package').version, '-v, --version')
    .usage('[options]')
    .option('--crawler <startURL>', 'Run Crawler, starting with URL startURL.')
    .option('--memory <startURL>', 'Run Memory, starting with URL startURL.')
    .option(
        '--psi <startURL>',
        'Run PageSpeed test for the home page of startURL.'
    )
    .parse(process.argv);

//

(async () => {
    const { crawler, memory, psi } = program;
    const startTS = Date.now();

    if (crawler) {
        console.log('Running Crawler test...');
        console.log(' ');
        const errors = await require('./crawler')(crawler, {
            debug: true,
            // maxCrawl: 100,
            cluster: {
                maxConcurrency: 5,
                monitor: true
            }
        });
        console.log('\n' + chalk.bgRedBright(` ERROR `));
        const dir = path.resolve(__dirname, '../../logs/crawler/');
        await fs.ensureDir(dir);
        await fs.writeJson(
            path.resolve(dir, `${encodeURIComponent(crawler)}.json`),
            parse(stringify(errors)),
            {
                spaces: '\t'
            }
        );
        Object.entries(errors).forEach(([type, errors]) => {
            console.log('❌ ' + chalk.underline(type));
            errors.forEach(err => {
                console.log(`● ${err.url}`);
                if (err.pageUrl) console.log(`  pageUrl: ${err.pageUrl}`);
                switch (type) {
                    case 'no gzip':
                    case 'large file': {
                        let size = err.contentLength;
                        if (size > 1024 * 1024) {
                            size = getSize(size);
                        } else if (size > 1024) {
                            size = getSize(size);
                        }
                        console.log(`  contentLength: ${size}`);
                        break;
                    }
                    case 'console error': {
                        console.log(`  ${err.message}`);
                        break;
                    }
                    default: {
                    }
                }
            });
        });
        // Object.keys(errors).forEach(type => {
        //     errors[type].forEach((err, index) => {
        //         const { message, res, type, ...infos } = err;
        //         const r = { ...infos };
        //         if (type === 'console error') r.message = message;
        //         errors[type][index] = r;
        //     });
        // });
        // Object.entries(errors).forEach(([key, value]) => {
        //     console.log(chalk.underline(key));
        //     console.log(JSON.stringify(value, null, 2));
        // });
        console.log('');
    } else if (memory) {
        console.log('Running Memory test for 60s...');
        console.log('');
        const result = await require('./memory')(memory, 60, true);
        result.forEach(r => {
            console.log(`● ${r.prevUrl}`);
            console.log(`⇢ ${r.newUrl}`);
            console.log(`   heap used:  ${getSize(r.JSHeapUsedSize)}`);
            console.log(`   heap total: ${getSize(r.JSHeapTotalSize)}`);
        });
        console.log('');
    } else if (psi) {
        console.log('Running PageSpeed tests...');
        console.log('');
        const result = await require('./psi')(psi);
        console.dir(result);
        console.log('');
    }

    const endTS = Date.now();
    console.log(`Elapsed ${((endTS - startTS) / 1000).toFixed(3)}ms`);

    // program.help();
})();

//

const getSize = size => {
    if (size > 1024 * 1024) {
        return (size / 1024 / 1024).toFixed(2) + 'MB';
    } else if (size > 1024) {
        return (size / 1024).toFixed(2) + 'KB';
    }
};
